src/validator/group.rs

```diff
@@ -0,0 +1,824 @@
+use super::types::ValidationState;
+use super::utilities::make_ede;
+use super::utilities::map_dname;
+use super::utilities::ttl_for_sig;
+use crate::base::cmp::CanonicalOrd;
+use crate::base::iana::class::Class;
+use crate::base::iana::ExtendedErrorCode;
+use crate::base::name::ToName;
+use crate::base::opt::exterr::ExtendedError;
+use crate::base::rdata::ComposeRecordData;
+use crate::base::Name;
+use crate::base::ParsedName;
+use crate::base::ParsedRecord;
+use crate::base::Record;
+use crate::base::Rtype;
+use crate::base::Ttl;
+use crate::dep::octseq::builder::with_infallible;
+use crate::dep::octseq::Octets;
+use crate::dep::octseq::OctetsFrom;
+use crate::net::client::request::RequestMessage;
+use crate::net::client::request::SendRequest;
+use crate::rdata::dnssec::Timestamp;
+use crate::rdata::AllRecordData;
+use crate::rdata::Dnskey;
+use crate::rdata::Rrsig;
+use crate::validate::RrsigExt;
+use bytes::Bytes;
+use moka::future::Cache;
+use ring::digest;
+use std::cmp::min;
+use std::fmt::Debug;
+use std::slice::Iter;
+use std::time::Duration;
+use std::vec::Vec;
+
+type RrType = Record<Name<Bytes>, AllRecordData<Bytes, ParsedName<Bytes>>>;
+type SigType = Record<Name<Bytes>, Rrsig<Bytes, Name<Bytes>>>;
+
+#[derive(Debug)]
+pub struct Group {
+    rr_set: Vec<RrType>,
+    sig_set: Vec<SigType>,
+}
+
+impl Group {
+    fn new(rr: ParsedRecord<'_, Bytes>) -> Result<Self, Error> {
+        let sig_record = match rr.to_record::<Rrsig<_, _>>()? {
+            None => {
+                return Ok(Self {
+                    rr_set: vec![to_bytes_record(&rr)?],
+                    sig_set: Vec::new(),
+                });
+            }
+            Some(record) => record,
+        };
+        let rrsig = sig_record.data();
+
+        let rrsig: Rrsig<Bytes, Name<Bytes>> =
+            Rrsig::<Bytes, Name<Bytes>>::new(
+                rrsig.type_covered(),
+                rrsig.algorithm(),
+                rrsig.labels(),
+                rrsig.original_ttl(),
+                rrsig.expiration(),
+                rrsig.inception(),
+                rrsig.key_tag(),
+                rrsig.signer_name().to_name::<Bytes>(),
+                Bytes::copy_from_slice(rrsig.signature().as_ref()),
+            )
+            .expect("should not fail");
+
+        let record: Record<Name<Bytes>, _> = Record::new(
+            sig_record.owner().to_name::<Bytes>(),
+            sig_record.class(),
+            sig_record.ttl(),
+            rrsig,
+        );
+
+        Ok(Self {
+            rr_set: Vec::new(),
+            sig_set: vec![record],
+        })
+    }
+
+    fn add(&mut self, rr: &ParsedRecord<'_, Bytes>) -> Result<(), ()> {
+        // First check owner.
+        if !self.rr_set.is_empty() {
+            if self.rr_set[0].owner() != &rr.owner().to_name::<Bytes>() {
+                return Err(());
+            }
+        } else if self.sig_set[0].owner().try_to_name::<Bytes>()
+            != rr.owner().try_to_name()
```

Comment:
Somehow this fails. It works if I replace try_to_name with to_name. Maybe @partim knows.

Reply:
If you have `Bytes`, (then `try_to_name` will never fail because the error type is `Infallible`), you should use `to_name`.

But in this case you probably should use `name_eq` like Terts suggested since this avoids actually assembling the name in newly allocated memory.

Reply:
I tried name_eq and it fails some of my tests

Reply:
That is worrisome. Should probably look into it.

Reply:
I'll keep this open and I'll try to find out what goes wrong.